source/extensions/filters/network/sni_dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -0,0 +1,75 @@
+#include "extensions/filters/network/sni_dynamic_forward_proxy/proxy_filter.h"
+
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/upstream/thread_local_cluster.h"
+
+#include "common/common/assert.h"
+#include "common/tcp_proxy/tcp_proxy.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SniDynamicForwardProxy {
+
+ProxyFilterConfig::ProxyFilterConfig(
+    const FilterConfig& proto_config,
+    Extensions::Common::DynamicForwardProxy::DnsCacheManagerFactory& cache_manager_factory,
+    Upstream::ClusterManager& cluster_manager)
+    : port_(static_cast<uint16_t>(proto_config.port_value())),
+      dns_cache_manager_(cache_manager_factory.get()),
+      dns_cache_(dns_cache_manager_->getCache(proto_config.dns_cache_config())),
+      cluster_manager_(cluster_manager) {}
+
+ProxyFilter::ProxyFilter(ProxyFilterConfigSharedPtr config) : config_(std::move(config)) {}
+
+using LoadDnsCacheEntryStatus = Common::DynamicForwardProxy::DnsCache::LoadDnsCacheEntryStatus;
+
+Network::FilterStatus ProxyFilter::onNewConnection() {
+  absl::string_view sni = read_callbacks_->connection().requestedServerName();
+  ENVOY_CONN_LOG(trace, "sni_dynamic_forward_proxy: new connection with server name '{}'",
+                 read_callbacks_->connection(), sni);
+
+  if (sni.empty()) {
+    return Network::FilterStatus::Continue;
+  }
+
+  // TODO(lizan): implement circuit breaker in SNI dynamic forward proxy
```

Comment:
Sorry what circuit breaker do we need here? I can't recall the details of what we do elsewhere. Can you be more detailed in the TODO?

Reply:
add a link to what is done in HTTP.